### DISCUSSION TITLE
Using Keyestudio CNC Shield V4

### DISCUSSION THREAD

**Discussion by VersusTheWorld89** [3DChameleon/3DChameleonMk4#16] [General]
I am trying to set up the code for my Keyestudio CNC Shield V4, thanks to the work of Liam Pulido. I am trying to use the code from https://www.3dchameleon.com/forum/main/comment/92bf7dfd-67ab-45d8-8d23-86de16afa65e?postId=6770a14a797c5c8b5ad99781, but that link goes back to the forum groups.

[Firmware.txt](https://github.com/user-attachments/files/20488700/Firmware.txt)

I dug deeper into the conversation and found the attached txt file. I deleted the default code on the Arduino and copied and pasted on the Arduino, but it gives me a lot of errors - below.

Thank you!

C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:136:1: error: 'SX1509' does not name a type
 SX1509 io;                        // Create an SX1509 object to be used throughout
 ^~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:161:1: error: 'Servo' does not name a type; did you mean 'Serial'?
 Servo filamentCutter;  // create servo object to control a servo
 ^~~~~
 Serial
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void setup()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:203:3: error: 'Wire' was not declared in this scope
   Wire.begin(); //start i2C
   ^~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:207:7: error: 'io' was not declared in this scope
   if (io.begin(SX1509_ADDRESS) == true)
       ^~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:218:3: error: 'oled' was not declared in this scope
   oled.begin(&Adafruit128x64, OLED_I2C_ADDRESS);
   ^~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:218:15: error: 'Adafruit128x64' was not declared in this scope
   oled.begin(&Adafruit128x64, OLED_I2C_ADDRESS);
               ^~~~~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:225:16: error: 'Adafruit5x7' was not declared in this scope
   oled.setFont(Adafruit5x7);
                ^~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void updateIOBlock()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:312:18: error: 'io' was not declared in this scope
       T0Loaded = io.digitalRead(SX1509_FILAMENT_0);
                  ^~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void displayText(int, String)':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:487:5: error: 'oled' was not declared in this scope
     oled.clear();
     ^~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void connectGillotine()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:740:3: error: 'filamentCutter' was not declared in this scope
   filamentCutter.attach(12);
   ^~~~~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:740:3: note: suggested alternative: 'filament'
   filamentCutter.attach(12);
   ^~~~~~~~~~~~~~
   filament
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void disconnectGillotine()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:746:3: error: 'filamentCutter' was not declared in this scope
   filamentCutter.detach();
   ^~~~~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:746:3: note: suggested alternative: 'filament'
   filamentCutter.detach();
   ^~~~~~~~~~~~~~
   filament
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void openGillotine()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:754:5: error: 'filamentCutter' was not declared in this scope
     filamentCutter.write(pos);              // tell servo to go to position in variable 'pos'
     ^~~~~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:754:5: note: suggested alternative: 'filament'
     filamentCutter.write(pos);              // tell servo to go to position in variable 'pos'
     ^~~~~~~~~~~~~~
     filament
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino: In function 'void closeGillotine()':
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:765:5: error: 'filamentCutter' was not declared in this scope
     filamentCutter.write(pos);              // tell servo to go to position in variable 'pos'
     ^~~~~~~~~~~~~~
C:\Users\kleak\Documents\Arduino\3DChameleon\3DChameleon.ino:765:5: note: suggested alternative: 'filament'
     filamentCutter.write(pos);              // tell servo to go to position in variable 'pos'
     ^~~~~~~~~~~~~~
     filament
exit status 1

Compilation error: 'SX1509' does not name a type

**Comment by wjsteele** [comment]
You need to add all the referenced libraries into your Arduino environment.  

Bill

**Reply by VersusTheWorld89** [reply]
Thanks Bill! However, this is all very new to me. I see how I can add libraries. Do you know where I can find the names of the referenced ones?

Kendall

**Comment by wjsteele** [comment]
The libraries are referenced in the code at the top... all the INCLUDE blocks... just add the library with the same name in it.

**Reply by VersusTheWorld89** [reply]
Thanks Bill.

For those who want to know what libraries I installed to stop the errors.
"Servo" by Michael Margolis
"SSD1306Ascii" By Bill Greiman
"SX1509 IO Expander" by SparkFun Electronics

**Comment by CoreXYForge** [comment]
Is your issue resolved? I was on vacation so I decided to put my 3Dchameleon stuff on hold. If you have any more problems, could you add a comment on my CNC Shield/generic MMU thread? I would prefer to have one megathread for this.

Liam

Edit: The link you have in your original post still has the "/forum/" part of the link. You will need this link (https://www.3dchameleon.com/old-forum/support/help-on-setup-for-cnc-shield-v4-ankermake-m5c) that has the "/old-forum/" in the URL

Edit 2: I assume you are Kendall Leake from the old forum (maybe you should censor the "\users\kleak")? I am surprised that your comment even went through since they said it would shut down like 2 weeks ago. I was just on vacation, so I was taking a break from 3Dprinting stuff. I am happy to help with any other issues you have.

**Reply by VersusTheWorld89** [reply]
Hey Liam,
Yeah, I can keep these posts related to the CNC Shield thread on the original forum. I just assumed we would not want that if it were going to go away eventually. 

I was able to get the errors to stop after installing the libraries below.
"Servo" by Michael Margolis
"SSD1306Ascii" By Bill Greiman
"SX1509 IO Expander" by SparkFun Electronics

I commented on the original post with this looking for more help :)

"I have been staring at this thread off and on for a while now. Lots of information! I hope you are still around even after the Github move.

I am trying to test Keystudio board after programming the board. I have the steppers, the Nano board and the switch attached to the Keystudio and the light on the Nano lights then slowly fades, is that normal? I press and hold the switch and nothing happens."

**Reply by CoreXYForge** [reply]
Sorry if I was unclear. I meant to keep it on this (https://github.com/3DChameleon/3DChameleonMk4/discussions/13) github discussion. I saw the comment on the original forum, but for some reason it the reply button just never appears.

**Reply by VersusTheWorld89** [reply]
Ok cool! Thanks for the link you want these discussions to take place on. I will post my question(s) there.

**Comment by Macabane** [comment]
Hi I am a complete nob so sorry if this is a dumb question. I am trying to get the 3Dchameleon firmware onto keyestudio CNC Sheild V4 Arduino Nano with A4988 drivers when i try to verify the code is get the following error. 

/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `displayText(int, String) [clone .constprop.14]':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:480: undefined reference to `SSD1306Ascii::clear()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `closeGillotine()':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:760: undefined reference to `Servo::write(int)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `openGillotine()':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:748: undefined reference to `Servo::write(int)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `connectGillotine()':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:736: undefined reference to `Servo::attach(int)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `processMoves()':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:537: undefined reference to `Servo::detach()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `Print::write(unsigned char const*, unsigned int)':
/usr/share/arduino/hardware/arduino/avr/cores/arduino/Print.cpp:38: undefined reference to `SX1509::SX1509()'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/Print.cpp:42: undefined reference to `Servo::Servo()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `pinMode':
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring_digital.c:30: undefined reference to `SX1509::begin(unsigned char, TwoWire&, unsigned char)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring_digital.c:30: undefined reference to `SX1509::pinMode(unsigned char, unsigned char, unsigned char)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring_digital.c:30: undefined reference to `SX1509::pinMode(unsigned char, unsigned char, unsigned char)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring_digital.c:30: undefined reference to `SX1509::pinMode(unsigned char, unsigned char, unsigned char)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring_digital.c:30: undefined reference to `SX1509::pinMode(unsigned char, unsigned char, unsigned char)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `micros':
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring.c:79: undefined reference to `SX1509::pinMode(unsigned char, unsigned char, unsigned char)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring.c:79: undefined reference to `SSD1306Ascii::init(DevType const*)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring.c:79: undefined reference to `SSD1306Ascii::setFont(unsigned char const*)'
/usr/share/arduino/hardware/arduino/avr/cores/arduino/wiring.c:79: undefined reference to `SSD1306Ascii::clear()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `twi_transmit':
/usr/share/arduino/hardware/arduino/avr/libraries/Wire/src/utility/twi.c:343: undefined reference to `Servo::detach()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `TwoWire::endTransmission() [clone .constprop.2]':
/usr/share/arduino/hardware/arduino/avr/libraries/Wire/src/Wire.cpp:232: undefined reference to `SX1509::digitalRead(unsigned char)'
/usr/share/arduino/hardware/arduino/avr/libraries/Wire/src/Wire.cpp:232: undefined reference to `SX1509::digitalRead(unsigned char)'
/usr/share/arduino/hardware/arduino/avr/libraries/Wire/src/Wire.cpp:232: undefined reference to `SX1509::digitalRead(unsigned char)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `TwoWire::peek()':
/usr/share/arduino/hardware/arduino/avr/libraries/Wire/src/Wire.cpp:308: undefined reference to `SX1509::digitalRead(unsigned char)'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `homeSelector()':
/home/eric/Arduino/libraries/sketch_jul06a/sketch_jul06a.ino:770: undefined reference to `Servo::detach()'
/tmp/cc0VVKQC.ltrans0.ltrans.o: In function `Print::write(unsigned char const*, unsigned int)':
/usr/share/arduino/hardware/arduino/avr/cores/arduino/Print.cpp:35: undefined reference to `SSD1306Ascii::write(unsigned char)'
collect2: error: ld returned 1 exit status

**Reply by CoreXYForge** [reply]
You most likely did not install these libraries:

"Servo" by Michael Margolis
"SSD1306Ascii" By Bill Greiman
"SX1509 IO Expander" by SparkFun Electronics

After you install those, it should upload and work fine :)

**Reply by Macabane** [reply]
Thank you for humoring a numb like me you were 100% right I had not loaded the repositories right and that was the problem 
for any other numbs like me i found this useful https://www.youtube.com/watch?v=YwVYe3hIoa0 
Thank you again

**Comment by Macabane** [comment]
[sketch_jul06a (copy).txt](https://github.com/user-attachments/files/21087264/sketch_jul06a.copy.txt)
[error.odt](https://github.com/user-attachments/files/21087266/error.odt)
added text versions of the firmware and error
